Cabal/Distribution/Simple/InstallDirs.hs

```diff
@@ -181,7 +183,9 @@ defaultInstallDirs comp userInstall _hasLibs = do
            JHC    -> "$compiler"
            LHC    -> "$compiler"
            UHC    -> "$pkgid"
-           _other -> "$abi" </> "$libname",
+           _other -> case buildOS of
+                        OSX      -> "$abi" -- OSX libs go into a single directory
```

Comment:
Is there any particular reason to make this conditional on the operating system? Placing dynamic libraries in per-package directories is actually a rather severe performance problem in general since the dynamic linker (on Linux at least) performs a linear scan through `RPATH` entries when looking for each library. For this reason just running `ghc --help` on Linux performs over 750 `open` system calls. This was first documented in [GHC #11587](https://ghc.haskell.org/trac/ghc/ticket/11587) and the obvious solution is to move all dynamic libraries into a single directory, just as you've done here.